core-compose/src/main/java/com/squareup/workflow/ui/compose/ComposeViewFactoryRoot.kt

```diff
@@ -56,6 +65,27 @@ fun ComposeViewFactoryRoot(
   @Composable override fun wrap(content: @Composable() () -> Unit) = wrapper(content)
 }
 
+/**
+ * Adds [content] to the composition, ensuring that it has been wrapped by a
+ * [ComposeViewFactoryRoot] if present in the [ViewEnvironment].
+ */
+@Composable internal fun wrapWithRootIfNecessary(
+  viewEnvironment: ViewEnvironment,
+  content: @Composable() () -> Unit
+) {
+  if (HasViewFactoryRootBeenApplied.current) {
```

Comment:
I'm missing where this gets set to true.

Reply:
Great question, I'll add a comment to the source. It gets "set" to true on line 79 below. This function can appear multiple times, at multiple nesting levels, inside the composition. The ambient has a default value of false, which means if this function is the top-most one in its subtree, it will read the value as false and drop down to the `else` case. That case will wrap the child Composable with this ambient providing the value true, so any nested occurrences of this function below this point will see that true value and not re-apply the wrapper.

Reply:
God, it was staring me right in the face.

Reply:
It's weird to look at if you're not thinking about recursion. I added a bunch of comments that I hope should clear this up.